ocaml/xapi-cli-server/cli_operations.ml

```diff
@@ -7684,19 +7644,91 @@ let update_resync_host _printer rpc session_id params =
   let host = Client.Host.get_by_uuid ~rpc ~session_id ~uuid in
   Client.Pool_update.resync_host ~rpc ~session_id ~host
 
-let host_apply_updates _printer rpc session_id params =
+let get_avail_updates_uri ~session_id ~task ~host =
+  let query =
+    [
+      ("session_id", [Ref.string_of session_id])
+    ; ("task_id", [Ref.string_of task])
+    ; ("host_refs", [Ref.string_of host])
+    ]
+  in
+  Uri.make ~path:Constants.get_updates_uri ~query () |> Uri.to_string
+
+let command_in_task ~rpc ~session_id ~fd ~host ~label f =
+  let task =
+    Client.Task.create ~rpc ~session_id
+      ~label:(Printf.sprintf "%s for host (ref=%s)" label (Ref.string_of host))
+      ~description:""
+  in
+  Client.Task.set_progress ~rpc ~session_id ~self:task ~value:(-1.0) ;
+  let command = f session_id task host in
+  finally
+    (fun () ->
+      marshal fd (Command command) ;
+      let response = ref (Response Wait) in
+      while !response = Response Wait do
+        response := unmarshal fd
+      done ;
+      let ok =
+        match !response with
+        | Response OK ->
+            true
+        | Response Failed ->
+            (* Need to check whether the thin cli managed to contact the server
+             * or not. If not, we need to mark the task as failed.
+             *)
+            if Client.Task.get_progress ~rpc ~session_id ~self:task < 0.0 then
+              Client.Task.set_status ~rpc ~session_id ~self:task ~value:`failure ;
+            false
+        | _ ->
+            false
+      in
+      wait_for_task_complete rpc session_id task ;
+      check_task_status ~rpc ~session_id ~task ~fd ~label ~ok
+        ~quiet_on_success:true ()
+    )
+    (fun () -> Client.Task.destroy ~rpc ~session_id ~self:task)
+
+let print_avail_updates ~rpc ~session_id ~fd ~host =
+  command_in_task ~rpc ~session_id ~fd ~host ~label:"Print available updates"
+    (fun session_id task host ->
+      PrintHttpGetJson (get_avail_updates_uri ~session_id ~task ~host)
+  )
+
+let print_update_guidance ~rpc ~session_id ~fd ~host =
+  command_in_task ~rpc ~session_id ~fd ~host ~label:"Print update guidance"
+    (fun session_id task host ->
+      PrintUpdateGuidance (get_avail_updates_uri ~session_id ~task ~host)
+  )
+
+let host_apply_updates fd printer rpc session_id params =
   let hash = List.assoc "hash" params in
-  ignore
-    (do_host_op rpc session_id ~multiple:false
-       (fun _ host ->
-         let host = host.getref () in
-         Client.Host.apply_updates ~rpc ~session_id ~self:host ~hash
-         |> List.iter (fun l ->
-                _printer (Cli_printer.PMsg (String.concat "; " l))
-            )
-       )
-       params ["hash"]
+  do_host_op rpc session_id ~multiple:false
+    (fun _ host ->
+      let host = host.getref () in
+      printer (Cli_printer.PMsg "Guidance of updates:") ;
+      print_update_guidance ~rpc ~session_id ~fd ~host ;
```

Comment:
Or print the guidance after applying updates?

Reply:
The working pattern of`xe` CLI and xapi is:
1.`xe` CLI connect to xapi CLI handler
2. xapi CLI handler sends a `command` back to `xe` CLI 
3. in the `command` handler, `xe` CLI issues a XAPI call according to the `command`.

The `command` handler in `xe` CLI can't remember the state of the previoius commands. And after applying updates, the guidance from the available updates will be cleaned up in xapi.